### ISSUE TITLE
Unable to update root span attributes

### ISSUE THREAD

**Issue by JohnDoneth** [open-telemetry/opentelemetry-erlang#570]
When working with DataDog and OpenTelemetry they expect certain attributes to be on the root span in order for the trace to show up in their error reporting segment of their APM application rather than just a trace with some errors contained. 

I don't see a way to get a span context for the root span so I can update the root span attributes. Due to how `opentelemetry_phoenix` creates the root spans for us we can't keep track of the root span context during creation so it would be ideal if we could access the currently bound root span or somehow walk up parent spans for the current span.

Another reason this could be helpful is to add arbitrary attributes during the runtime of the application to our root Phoenix span such as the request ID, current user ID, etc, that aren't currently populated by the `opentelemetry_phoenix` library. As of the moment, those attributes would be hidden on a subspan which would be trickier to find and view.

**Comment by tsloughter** [comment]
When are you trying to set the attributes? I would think that it is your root span you have access to in your plug or controller, or is it that the root span is from Cowboy?

Are all the attributes you want to set from the headers? While you should be able to do what you are asking, we also need support for adding headers as attributes on start.

**Comment by JohnDoneth** [comment]
> I would think that it is your root span you have access to in your plug or controller, or is it that the root span is from Cowboy?

That's true if we don't add another span for the plug itself then I could add the attributes to the current span in a plug which would also be the root span for Phoenix. I think that should work for that use-case!

Unfortunately that doesn't solve the other concern with being able to set the root span attributes necessary for DataDog to mark the trace as an error. Described [here](https://docs.datadoghq.com/tracing/error_tracking/#use-span-tags-to-track-error-spans) which requires `error.stack`, `error.message`, and `error.type` to be set on the root span. Due to the nature of some of our errors we'd want to be able to set this at any point in the program which could occur in a deeply nested span so that DataDog recognizes them as traces that should show up in their error tracker.

**Comment by bryannaegele** [comment]
Most things can be set within a plug or the controller as noted. Based on the link you posted and error tracking, it sounds like this is an issue with Datadog failing to honor the standard. This kind of thing is not new with them but unfortunately isn't something that's an otel issue. I'm surprised they require this behavior.

The only thing I can suggest is you could set up some kind of registry in your code where you can track root spans, but what they seem to be requiring here is against the otel specification.

**Comment by tsloughter** [comment]
Had a thought. I don't know enough about plug/phoenix to say how this might work but I was thinking about how I might handle similar in Elli.

If there is a way to setup a handler for when there is an error case then all you'd need is to keep the SpanCtx of the root span in the metadata (I guess `conn`?).

Then in the handler you grab that SpanCtx from there and set the attributes.

So basically, track the root in `conn`?

**Comment by tsloughter** [comment]
And yea, as @bryannaegele mentions it'd be nice if Datadog supported https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md#attributes -- which the instrumentation libraries do set. But obviously that is a wish and not a solution :), hopefully we can make it work anyway.

**Comment by bryannaegele** [comment]
Yeah but the conn isn't passed anywhere past the controller handler so you couldn't grab it from elsewhere. You could stick it in pdict but then if you cross processes then same issue.

The other thing here is no other instrumentation library is going to support this. An ecto error isn't going to be able to be grabbed and set on some root span, so this is kind of a non-starter unfortunately.

**Comment by tsloughter** [comment]
@JohnDoneth oh, just realized also, are you using the OTLP exporter to the Collector and then having that export to Datadog? Or are you sending with the Zipkin exporter?

**Comment by tsloughter** [comment]
@bryannaegele my assumption was that any error had to get back up to the top level to return a proper response, so at that point it is available with `conn`.

**Comment by JohnDoneth** [comment]
@tsloughter we're using the OpenTelemetry collector with its DataDog exporter.

**Comment by tsloughter** [comment]
@bryannaegele I mean, that must be how it is setting the error attributes already, right?

**Comment by bryannaegele** [comment]
I mean, if it's uncaught and a 500 is returned, the error would already be getting set. If the code is just returning error tuples and you decide to return a different response code then any exception info isn't going to be included unless you're passing exception structs around. But again, let's say you're making an http call out and get a 500 or retry for success, that call is errored but can still result in a successful operation in general..

**Comment by bryannaegele** [comment]
Also, we may be conflating errored spans and exceptions. Not all errored spans are going to be recorded exceptions and have stacktraces, etc

**Comment by tsloughter** [comment]
But any errored span could be made to throw an exception. I was thinking how in Elli I have middlewares and there are handlers for exceptions thrown with the type and stacktrace, so setting datadog specific attributes in a custom middleware would be doable. It might mean re-raising an exception I caught deeper down to make this happen, but its at least an option -- 
since `error.stack` is required an exception must be thrown somewhere as its the only way to get a stacktrace in Erlang anymore.

And if it is a separate process you can still get this through the reason and stacktrace on an `EXIT` signal.

Passing down the `conn` with the root SpanCtx as metadata might make it neater in some cases though, if possible.

**Comment by tsloughter** [comment]
Not sure if there was any satisfactory resolution here? Is there more to discuss? I've lost the context from my head but I believe exception and error on spans keeps coming up, maybe this can be closed and consolidated into another issue?

**Comment by tomtaylor** [comment]
I also wanted to have a handle on the root span, so I can attach attributes to it from anywhere to support 'wide event' style telemetry. I ended up making a telemetry handler that runs after the OpenTelemetry handlers for all the key transaction starts. It gets a handle on the span context and stashes it in the process. Then it uses ProcessTree to find the span context from the same process or any child processes. For example:

```elixir
defmodule Example.OpenTelemetry.RootSpan do
  @moduledoc """
  This module provides functions for stashing the root span context in the
  process dictionary, and setting attributes on the root span from anywhere in
  the code.
  """

  # The process key under which we stash the root span context.
  @key :otel_root_span

  @doc """
  Attach telemetry handlers to events that signal the start of a new
  OpenTelemetry trace. This allows us to stash the root span context in the
  process dictionary, so that we can set attributes on the root span from
  anywhere in the code, without needing to pass around the span context.
  """
  def attach_telemetry_handlers() do
    :ok =
      :telemetry.attach_many(
        "otel-root-span",
        [
          [:phoenix, :endpoint, :start],
          [:cowboy, :request, :start]
        ],
        &__MODULE__.handle_event/4,
        nil
      )
  end

  @doc false
  def handle_event(_event, _measurements, _metadata, _config) do
    # At the beginning of a span, we stash the span context for the root span
    # into the process dictionary, so that we can access it later when we want
    # to set attributes on the root span.
    stash_span_ctx()
  end

  @doc """
  Stash the current span context in the process dictionary, if it's not already
  stashed. This is called at the beginning of a span when the current span ctx
  is the root span.
  """
  def stash_span_ctx() do
    case ProcessTree.get(@key, cache: false) do
      nil ->
        ctx = OpenTelemetry.Tracer.current_span_ctx()
        Process.put(@key, ctx)
        :ok

      _ctx ->
        :ok
    end
  end

  @doc """
  Set an attribute on the root span of the current trace.

  Returns `false` if there is no root span context stashed in the process
  dictionary. Otherwise, returns the result of
  `OpenTelemetry.Span.set_attribute/3`.
  """
  def set_attribute(key, value) do
    case ProcessTree.get(@key) do
      nil -> false
      ctx -> OpenTelemetry.Span.set_attribute(ctx, key, value)
    end
  end

  @doc """
  See `set_attribute/2`.
  """
  def set_attributes(attributes) do
    case ProcessTree.get(@key) do
      nil -> false
      ctx -> OpenTelemetry.Span.set_attributes(ctx, attributes)
    end
  end
end
```
